### PR TITLE
add tag resource

### PR DESCRIPTION
**Pull request by davidbloss**
## TopHatting

<details>
<summary>Step 1: Wrote Terraform file</summary>
<br>
 <pre><code>
data "opslevel_repository" "foo" {
  alias = "github.com:example-test/example"
}

data "opslevel_service" "serve_foo" {
  alias = "a_test_service"
}

resource "opslevel_tag" "repo-id" {
  resource_type = "Repository"
  resource_id   = data.opslevel_repository.foo.id

  key   = "type2"
  value = "frontend1-asdf"
}

resource "opslevel_tag" "repo-alias" {
  resource_type = "Repository"
  resource_id   = "github.com:example-test2/example-2"

  key   = "test-test"
  value = "breaker-breaker-asdf"
}

resource "opslevel_tag" "foo_domain" {
  resource_type = "Domain"
  resource_id   = "Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzUyNDg0Nw"

  key   = "dmn-cheesy"
  value = "pasta-asdf"
}

resource "opslevel_tag" "foo_service" {
  resource_type = "Service"
  resource_id   = data.opslevel_service.serve_foo.id

  key   = "svc-cheesy"
  value = "attack-asdf"
}

resource "opslevel_tag" "foo_system" {
  resource_type = "System"
  resource_id   = "test"

  key   = "sys-cheesy"
  value = "attack-asdf"
}

resource "opslevel_tag" "foo_team" {
  resource_type = "Team"
  resource_id   = "platform"

  key   = "team-cheesy"
  value = "attack-asdf"
}
</code> </pre>
</details>

<details>
<summary>Step 2: terraform apply</summary>
<br>
 <pre><code> 
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  \# opslevel_tag.foo_domain will be created
  + resource "opslevel_tag" "foo_domain" {
      + id                = (known after apply)
      + key               = "dmn-cheesy"
      + last_updated      = (known after apply)
      + resource_id       = "Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzUyNDg0Nw"
      + resource_type     = "Domain"
      + returned_resource = (known after apply)
      + value             = "pasta-again"
    }

  \# opslevel_tag.foo_service will be created
  + resource "opslevel_tag" "foo_service" {
      + id                = (known after apply)
      + key               = "svc-cheesy"
      + last_updated      = (known after apply)
      + resource_id       = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS81Nzk3"
      + resource_type     = "Service"
      + returned_resource = (known after apply)
      + value             = "attack-again"
    }

  \# opslevel_tag.foo_system will be created
  + resource "opslevel_tag" "foo_system" {
      + id                = (known after apply)
      + key               = "sys-cheesy"
      + last_updated      = (known after apply)
      + resource_id       = "test"
      + resource_type     = "System"
      + returned_resource = (known after apply)
      + value             = "attack-again"
    }

  \# opslevel_tag.foo_team will be created
  + resource "opslevel_tag" "foo_team" {
      + id                = (known after apply)
      + key               = "team-cheesy"
      + last_updated      = (known after apply)
      + resource_id       = "platform"
      + resource_type     = "Team"
      + returned_resource = (known after apply)
      + value             = "attack-again"
    }

  \# opslevel_tag.repo-alias will be created
  + resource "opslevel_tag" "repo-alias" {
      + id                = (known after apply)
      + key               = "test-test"
      + last_updated      = (known after apply)
      + resource_id       = "github.com:suraj-testing2/Kitty_Ice_cream_cone"
      + resource_type     = "Repository"
      + returned_resource = (known after apply)
      + value             = "breaker-breaker"
    }

  \# opslevel_tag.repo-id will be created
  + resource "opslevel_tag" "repo-id" {
      + id                = (known after apply)
      + key               = "type2"
      + last_updated      = (known after apply)
      + resource_id       = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzMw"
      + resource_type     = "Repository"
      + returned_resource = (known after apply)
      + value             = "frontend1"
    }

Plan: 6 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

opslevel_tag.foo_system: Creating...
opslevel_tag.repo-id: Creating...
opslevel_tag.foo_service: Creating...
opslevel_tag.foo_domain: Creating...
opslevel_tag.foo_team: Creating...
opslevel_tag.repo-alias: Creating...
opslevel_tag.repo-alias: Creation complete after 2s [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUxNDc5]
opslevel_tag.foo_system: Creation complete after 2s [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg2]
opslevel_tag.foo_domain: Creation complete after 2s [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg3]
opslevel_tag.foo_team: Creation complete after 2s [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg1]
opslevel_tag.repo-id: Creation complete after 2s [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUxNDgy]
opslevel_tag.foo_service: Creation complete after 2s [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTkw]

Apply complete! Resources: 6 added, 0 changed, 0 destroyed.
</code> </pre>
</details>

<details>
<summary>Step 3: terraform update - after modifying file above</summary>
<pre><code>Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  \# opslevel_tag.foo_domain will be updated in-place
  ~ resource "opslevel_tag" "foo_domain" {
        id                = "Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg3"
      ~ value             = "pasta-again" -> "pasta-asdf"
        # (4 unchanged attributes hidden)
    }

  \# opslevel_tag.foo_service will be updated in-place
  ~ resource "opslevel_tag" "foo_service" {
        id                = "Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTkw"
      ~ value             = "attack-again" -> "attack-asdf"
        # (4 unchanged attributes hidden)
    }

  \# opslevel_tag.foo_system will be updated in-place
  ~ resource "opslevel_tag" "foo_system" {
        id                = "Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg2"
      ~ value             = "attack-again" -> "attack-asdf"
        # (4 unchanged attributes hidden)
    }

  \# opslevel_tag.foo_team will be updated in-place
  ~ resource "opslevel_tag" "foo_team" {
        id                = "Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg1"
      ~ value             = "attack-again" -> "attack-asdf"
        # (4 unchanged attributes hidden)
    }

  \# opslevel_tag.repo-alias will be updated in-place
  ~ resource "opslevel_tag" "repo-alias" {
        id                = "Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUxNDc5"
      ~ value             = "breaker" -> "breaker-breaker-asdf"
        # (5 unchanged attributes hidden)
    }

  \# opslevel_tag.repo-id will be updated in-place
  ~ resource "opslevel_tag" "repo-id" {
        id                = "Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUxNDgy"
      ~ value             = "frontend123" -> "frontend1-asdf"
        # (5 unchanged attributes hidden)
    }

Plan: 0 to add, 6 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

opslevel_tag.foo_system: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg2]
opslevel_tag.repo-id: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUxNDgy]
opslevel_tag.repo-alias: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUxNDc5]
opslevel_tag.foo_team: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg1]
opslevel_tag.foo_domain: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg3]
opslevel_tag.foo_service: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTkw]
opslevel_tag.foo_system: Modifications complete after 2s [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg2]
opslevel_tag.repo-id: Modifications complete after 2s [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUxNDgy]
opslevel_tag.foo_service: Modifications complete after 2s [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTkw]
opslevel_tag.foo_team: Modifications complete after 2s [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg1]
opslevel_tag.repo-alias: Modifications complete after 3s [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUxNDc5]
opslevel_tag.foo_domain: Modifications complete after 3s [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg3]

Apply complete! Resources: 0 added, 6 changed, 0 destroyed.
</code> </pre>
</details>

<details>
<summary>Step 4: terraform destroy</summary>
<br>
<pre><code> 
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  \# opslevel_tag.foo_domain will be destroyed
  - resource "opslevel_tag" "foo_domain" {
      - id                = "Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg3" -> null
      - key               = "dmn-cheesy" -> null
      - last_updated      = "Tuesday, 26-Sep-23 16:10:40 CDT" -> null
      - resource_id       = "Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzUyNDg0Nw" -> null
      - resource_type     = "Domain" -> null
      - returned_resource = "Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzUyNDg0Nw" -> null
      - value             = "pasta-asdf" -> null
    }

  \# opslevel_tag.foo_service will be destroyed
  - resource "opslevel_tag" "foo_service" {
      - id                = "Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTkw" -> null
      - key               = "svc-cheesy" -> null
      - last_updated      = "Tuesday, 26-Sep-23 16:10:38 CDT" -> null
      - resource_id       = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS81Nzk3" -> null
      - resource_type     = "Service" -> null
      - returned_resource = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS81Nzk3" -> null
      - value             = "attack-asdf" -> null
    }

  \# opslevel_tag.foo_system will be destroyed
  - resource "opslevel_tag" "foo_system" {
      - id                = "Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg2" -> null
      - key               = "sys-cheesy" -> null
      - last_updated      = "Tuesday, 26-Sep-23 16:10:38 CDT" -> null
      - resource_id       = "test" -> null
      - resource_type     = "System" -> null
      - returned_resource = "Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzExOTc" -> null
      - value             = "attack-asdf" -> null
    }

  \# opslevel_tag.foo_team will be destroyed
  - resource "opslevel_tag" "foo_team" {
      - id                = "Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg1" -> null
      - key               = "team-cheesy" -> null
      - last_updated      = "Tuesday, 26-Sep-23 16:10:39 CDT" -> null
      - resource_id       = "platform" -> null
      - resource_type     = "Team" -> null
      - returned_resource = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS85NzU5" -> null
      - value             = "attack-asdf" -> null
    }

  \# opslevel_tag.repo-alias will be destroyed
  - resource "opslevel_tag" "repo-alias" {
      - id                = "Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUxNDc5" -> null
      - key               = "test-test" -> null
      - last_updated      = "Tuesday, 26-Sep-23 16:10:38 CDT" -> null
      - resource_id       = "github.com:suraj-testing2/Kitty_Ice_cream_cone" -> null
      - resource_type     = "Repository" -> null
      - returned_resource = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDY5OQ" -> null
      - value             = "breaker-breaker-asdf" -> null
    }

  \# opslevel_tag.repo-id will be destroyed
  - resource "opslevel_tag" "repo-id" {
      - id                = "Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUxNDgy" -> null
      - key               = "type2" -> null
      - last_updated      = "Tuesday, 26-Sep-23 16:10:38 CDT" -> null
      - resource_id       = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzMw" -> null
      - resource_type     = "Repository" -> null
      - returned_resource = "Z2lkOi8vb3BzbGV2ZWwvUmVwb3NpdG9yaWVzOjpHaXRodWIvNDYzMw" -> null
      - value             = "frontend1-asdf" -> null
    }

Plan: 0 to add, 0 to change, 6 to destroy.

Do you really want to destroy all resources?
  Terraform will destroy all your managed infrastructure, as shown above.
  There is no undo. Only 'yes' will be accepted to confirm.

  Enter a value: yes

opslevel_tag.foo_team: Destroying... [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg1]
opslevel_tag.repo-alias: Destroying... [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUxNDc5]
opslevel_tag.foo_service: Destroying... [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTkw]
opslevel_tag.repo-id: Destroying... [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUxNDgy]
opslevel_tag.foo_domain: Destroying... [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg3]
opslevel_tag.foo_system: Destroying... [id=Z2lkOi8vb3BzbGV2ZWwvVGFnLzYwNTUyOTg2]
opslevel_tag.repo-alias: Destruction complete after 0s
opslevel_tag.repo-id: Destruction complete after 0s
opslevel_tag.foo_team: Destruction complete after 0s
opslevel_tag.foo_service: Destruction complete after 0s
opslevel_tag.foo_system: Destruction complete after 1s
opslevel_tag.foo_domain: Destruction complete after 1s

Destroy complete! Resources: 6 destroyed.
</code> </pre>
</details>

### CODE REVIEW

File: examples/resources/opslevel_tag/resource.tf
Comment (taimoor-at-opslevel) [inline examples/resources/opslevel_tag/resource.tf:1]: Can we add an `import.sh` file in this directory for the tags resource?

**Review by taimoor-at-opslevel — COMMENTED** [review]
Can you add some tophatting output for the new `opslevel_tag` resource running on your machine? 

[example](https://github.com/OpsLevel/terraform-provider-opslevel/pull/103)

**Review by rocktavious — APPROVED** [review]
@davidbloss this looks good to me but i'd like to do some manual testing.  We can merge this if you want but lets hold off on releasing.

File: examples/resources/opslevel_tag/resource.tf
Comment (davidbloss) [reply examples/resources/opslevel_tag/resource.tf:1]: Last week @rocktavious mentioned something about us holding off on an `import.sh` for this one.
I think it had to do with us needing to define a custom `schema.ResourceImporter` since this is a special case?

**Comment by davidbloss** [comment]
> Can you add some tophatting output for the new `opslevel_tag` resource running on your machine?
> 
> [example](https://github.com/OpsLevel/terraform-provider-opslevel/pull/103)

Definitely! I put this off but have one on the way

File: examples/resources/opslevel_tag/resource.tf
Comment (rocktavious) [reply examples/resources/opslevel_tag/resource.tf:1]: Yes - this resource is not easily importable because the customer would have to use a stanza like

```sh
terraform import Service:Z2lkOi8vMTIzNDU2Nzg5OTg3NjU0MzIx:Z2lkOi8vMTkyODM3NDY1NTY0NzM4Mjkx
```

or

```sh
terraform import InfrastructureResource:Z2lkOi8vMTIzNDU2Nzg5OTg3NjU0MzIx:Z2lkOi8vMTkyODM3NDY1NTY0NzM4Mjkx
```

and then yes we'll have to write a customer importer to handle the 3 peices of information we need

- TaggableResource
- ResourceID
- TagID

Its non-trivial AND customers likely will not need to do this so i'd rather wait until it becomes a huge need.

File: opslevel/resource_opslevel_tag.go
Comment (rocktavious) [inline opslevel/resource_opslevel_tag.go:28]: this field should be required.

File: opslevel/resource_opslevel_tag.go
Comment (rocktavious) [inline opslevel/resource_opslevel_tag.go:34]: this field should be required

File: opslevel/resource_opslevel_tag.go
Comment (rocktavious) [inline opslevel/resource_opslevel_tag.go:41]: this field is required

File: opslevel/resource_opslevel_tag.go
Comment (rocktavious) [inline opslevel/resource_opslevel_tag.go:47]: .... should this field be required?  I guess what happens if the customer doesn't set this field.  Can we create a tag with an empty value?

File: opslevel/resource_opslevel_tag.go
Comment (rocktavious) [inline opslevel/resource_opslevel_tag.go:150]: i don't think we want to be updating this value.   Is there a case where you think the ID we asked for would be returned as a different ID that we need to update the state?  I think we can remove this.

File: opslevel/resource_opslevel_tag.go
Comment (rocktavious) [inline opslevel/resource_opslevel_tag.go:24]: since this takes an alias or an id - would `resource_identifier` be more apt of a name?

File: examples/resources/opslevel_tag/resource.tf
Comment (rocktavious) [inline examples/resources/opslevel_tag/resource.tf:51]: These are GREAT examples! 🎉

File: opslevel/resource_opslevel_tag.go
Comment (davidbloss) [reply opslevel/resource_opslevel_tag.go:24]: I think so 👍

File: opslevel/resource_opslevel_tag.go
Comment (davidbloss) [reply opslevel/resource_opslevel_tag.go:150]: No, it should be the same... I'll drop this

**Review by rocktavious — CHANGES_REQUESTED** [review]
A couple of changes and a few questions i'm not sure the answers too.

File: opslevel/resource_opslevel_tag.go
Comment (davidbloss) [reply opslevel/resource_opslevel_tag.go:47]: This should be required. I just tried to create a tag with an empty value and got an API Error saying `'value' Value is too short (minimum is 1 character)`

**Review by rocktavious — APPROVED** [review]
🎉

**Review by taimoor-at-opslevel — APPROVED** [review]
LGTM

**Pull request merged**
